File: evaluation/S0/python/Hard/2103-delete-duplicate-folders-in-system.py
# Source: https://leetcode.com/problems/delete-duplicate-folders-in-system/   |   Difficulty: Hard
#
# Problem Description:
# Due to a bug, there are many duplicate folders in a file system. You are given a 2D array paths, where paths[i] is an array representing an absolute path to the ith folder in the file system.
#
#
# 	For example, ["one", "two", "three"] represents the path "/one/two/three".
#
#
# Two folders (not necessarily on the same level) are identical if they contain the same non-empty set of identical subfolders and underlying subfolder structure. The folders do not need to be at the root level to be identical. If two or more folders are identical, then mark the folders as well as all their subfolders.
#
#
# 	For example, folders "/a" and "/b" in the file structure below are identical. They (as well as their subfolders) should all be marked:
#
# 	
# 		/a
# 		/a/x
# 		/a/x/y
# 		/a/z
# 		/b
# 		/b/x
# 		/b/x/y
# 		/b/z
# 	
# 	
# 	However, if the file structure also included the path "/b/w", then the folders "/a" and "/b" would not be identical. Note that "/a/x" and "/b/x" would still be considered identical even with the added folder.
#
#
# Once all the identical folders and their subfolders have been marked, the file system will delete all of them. The file system only runs the deletion once, so any folders that become identical after the initial deletion are not deleted.
#
# Return the 2D array ans containing the paths of the remaining folders after deleting all the marked folders. The paths may be returned in any order.
#
# Example:
# Input: paths = [["a"],["c"],["d"],["a","b"],["c","b"],["d","a"]]
# Output: [["d"],["d","a"]]
# Explanation: The file structure is as shown.
# Folders "/a" and "/c" (and their subfolders) are marked for deletion because they both contain an empty
# folder named "b".
#
# Constraints:
# 1 <= paths.length <= 2 * 104
# 	1 <= paths[i].length <= 500
# 	1 <= paths[i][j].length <= 10
# 	1 <= sum(paths[i][j].length) <= 2 * 105
# 	path[i][j] consists of lowercase English letters.
# 	No two paths lead to the same folder.
# 	For any folder not at the root level, its parent folder will also be in the input.
#

class Solution(object):
    def deleteDuplicateFolder(self, paths):
        """
        :type paths: List[List[str]]
        :rtype: List[List[str]]
        """
        from collections import defaultdict

        tree = lambda: defaultdict(tree)
        root = tree()
        for path in paths:
            cur = root
            for folder in path:
                cur = cur[folder]

        lookup = defaultdict(list)

        def serialize(node):
            if not node:
                return ""
            serials = []
            for folder in sorted(node):
                serials.append(folder + "(" + serialize(node[folder]) + ")")
            serial = "".join(serials)
            lookup[serial].append(node)
            return serial

        serialize(root)

        def collect(node, path, result):
            for folder in node:
                child = node[folder]
                serial = folder + "(" + serialize(child) + ")"
                if len(lookup[serial]) == 1:
                    result.append(path + [folder])
                    collect(child, path + [folder], result)

        result = []
        collect(root, [], result)
        return result

File: evaluation/S3/java/Hard/2103-delete-duplicate-folders-in-system.java
// Source: https://leetcode.com/problems/delete-duplicate-folders-in-system/   |   Difficulty: Hard
//
// Problem Description:
// Due to a bug, there are many duplicate folders in a file system. You are given a 2D array paths, where paths[i] is an array representing an absolute path to the ith folder in the file system.
//
//
// 	For example, ["one", "two", "three"] represents the path "/one/two/three".
//
//
// Two folders (not necessarily on the same level) are identical if they contain the same non-empty set of identical subfolders and underlying subfolder structure. The folders do not need to be at the root level to be identical. If two or more folders are identical, then mark the folders as well as all their subfolders.
//
//
// 	For example, folders "/a" and "/b" in the file structure below are identical. They (as well as their subfolders) should all be marked:
//
// 	
// 		/a
// 		/a/x
// 		/a/x/y
// 		/a/z
// 		/b
// 		/b/x
// 		/b/x/y
// 		/b/z
// 	
// 	
// 	However, if the file structure also included the path "/b/w", then the folders "/a" and "/b" would not be identical. Note that "/a/x" and "/b/x" would still be considered identical even with the added folder.
//
//
// Once all the identical folders and their subfolders have been marked, the file system will delete all of them. The file system only runs the deletion once, so any folders that become identical after the initial deletion are not deleted.
//
// Return the 2D array ans containing the paths of the remaining folders after deleting all the marked folders. The paths may be returned in any order.
//
// Example:
// Input: paths = [["a"],["c"],["d"],["a","b"],["c","b"],["d","a"]]
// Output: [["d"],["d","a"]]
// Explanation: The file structure is as shown.
// Folders "/a" and "/c" (and their subfolders) are marked for deletion because they both contain an empty
// folder named "b".
//
// Constraints:
// 1 <= paths.length <= 2 * 104
// 	1 <= paths[i].length <= 500
// 	1 <= paths[i][j].length <= 10
// 	1 <= sum(paths[i][j].length) <= 2 * 105
// 	path[i][j] consists of lowercase English letters.
// 	No two paths lead to the same folder.
// 	For any folder not at the root level, its parent folder will also be in the input.
//
// Helpful references (internal KB):
// - Strongly connected components and the condensation graph (graph, array, dfs, enumeration)
//   • When to use: Use this when systematically exploring all possible configurations, paths, or subsets within a graph or array. It's ideal for finding every solution or verifying properties across all potential states.
//   • Idea: Depth-First Search (DFS) is a graph traversal algorithm that explores as far as possible along each branch before backtracking. When combined with enumeration, it systematically generates all possible combinations or paths, typically with exponential time complexity O(V+E) for traversal or O(N!) / O(2^N) for enumeration.
//   • Invariants: The current path/subset being built is valid up to the current depth.; All previously explored branches from the current node have been fully processed.
//   • Tips: Define clear base cases for recursion termination.; Manage state changes carefully when entering and exiting recursive calls.
//   • Pitfalls: Infinite recursion due to missing or incorrect base cases.; Redundant computations if states are not properly managed or memoized.
// - Aho-Corasick Algorithm for Pattern Searching (string, array, aho-corasick, enumeration)
//   • When to use: Use when you need to find all occurrences of a fixed set of patterns (a dictionary) within a larger text efficiently. It is particularly effective for scenarios with many patterns or long texts where patterns might overlap.
//   • Idea: The Aho-Corasick algorithm constructs a finite automaton from a dictionary of patterns, then uses this automaton to search for all pattern occurrences in a text in a single pass. It achieves a time complexity of O(N + M + K) where N is text length, M is total pattern length, and K is total matches.
//   • Invariants: The trie structure correctly represents all prefixes of all patterns in the dictionary.; Each node's failure link points to the longest proper suffix of the string ending at that node which is also a prefix in the trie.
//   • Tips: Build the trie first from all patterns, marking terminal nodes.; Compute failure links using a BFS-like approach, similar to KMP's LPS array.
//   • Pitfalls: Incorrectly building failure links can lead to missed matches or infinite loops.; Forgetting to handle output links (dictionary links) will miss patterns found via failure transitions.
// - Suffix Automaton (string, enumeration)
//   • When to use: When you need to efficiently represent all substrings of a given string and perform various string operations like counting distinct substrings or pattern matching in linear time.
//   • Idea: A Suffix Automaton is a minimal finite automaton that accepts all suffixes of a given string. It implicitly represents all substrings and can be constructed in linear time, O(N) where N is the string length.
//   • Invariants: Every path from the initial state spells a substring of the original string.; Each state represents a set of substrings that share the same end positions in the original string.
//   • Tips: Understand the concept of `endpos` sets and how they define states.; The number of states and transitions in a suffix automaton is linear with respect to string length.
//   • Pitfalls: The construction algorithm can be complex and prone to off-by-one errors or incorrect link updates.; Debugging state transitions and suffix links requires careful attention.
//
// Notes:
// - Keep the given method signatures unchanged.
// - Use only Java standard library (`java.*`); no external dependencies.
// - Do not change input/output formats or return value semantics.
// - Conformance: single public class 'Solution'; no 'package' statement; place any extra types as static inner classes.
// - If you introduce any helper functions, include their full implementations in this file; no hidden helpers.

class Solution {
    public List<List<String>> deleteDuplicateFolder(List<List<String>> paths) {
        // Step 1: Build the folder structure as a tree
        TrieNode root = new TrieNode();
        for (List<String> path : paths) {
            TrieNode current = root;
            for (String folder : path) {
                current.children.putIfAbsent(folder, new TrieNode());
                current = current.children.get(folder);
            }
        }

        // Step 2: Serialize the subtree structures and count occurrences
        Map<String, Integer> subtreeCount = new HashMap<>();
        serializeSubtree(root, subtreeCount);

        // Step 3: Collect paths of remaining folders
        List<List<String>> result = new ArrayList<>();
        collectRemainingPaths(root, new ArrayList<>(), subtreeCount, result);
        return result;          
    }
}
